arangod/Replication2/Storage/WAL/LogPersistor.cpp

```diff
@@ -0,0 +1,315 @@
+
+    auto* payload = entry.logPayload();
+    TRI_ASSERT(payload != nullptr);
+    header.payloadSize = payload->byteSize();
+    _buffer.append(Record::CompressedHeader{header});
+    _buffer.append(payload->slice().getDataPtr(), payload->byteSize());
+    return header.payloadSize;
+  }
+
+  std::uint64_t writeMetaEntry(LogEntry const& entry) {
+    TRI_ASSERT(entry.hasMeta());
+    Record::Header header;
+    header.index = entry.logIndex().value;
+    header.term = entry.logTerm().value;
+    header.type = RecordType::wMeta;
+    header.payloadSize = 0;  // we initialize this to zero so the compiler is
+                             // happy, but will update it later
+
+    // the meta entry is directly encoded into the buffer, so we have to write
+    // the header beforehand and update the size afterwards
+    _buffer.append(Record::CompressedHeader{header});
+    auto pos = _buffer.size() - sizeof(Record::CompressedHeader::payloadSize);
+
+    auto* payload = entry.meta();
+    TRI_ASSERT(payload != nullptr);
+    VPackBuilder builder(_buffer.buffer());
+    payload->toVelocyPack(builder);
+    header.payloadSize =
+        _buffer.size() - pos - sizeof(Record::CompressedHeader::payloadSize);
+
+    // reset to the saved position so we can write the actual size
+    _buffer.resetTo(pos);
+    _buffer.append(static_cast<decltype(Record::CompressedHeader::payloadSize)>(
+        header.payloadSize));
+    _buffer.advance(header.payloadSize);
+    return header.payloadSize;
+  }
+
+  void writePaddingBytes(
+      std::uint32_t payloadSize) {  // write zeroed out padding bytes
+    auto numPaddingBytes = Record::paddedPayloadSize(payloadSize) - payloadSize;
+    TRI_ASSERT(numPaddingBytes < 8);
+    std::uint64_t const zero = 0;
+    _buffer.append(&zero, numPaddingBytes);
+  }
+
+  void writeFooter(std::size_t startPos) {
+    auto size = _buffer.size() - startPos;
+    Record::Footer footer{
+        .crc32 = static_cast<std::uint32_t>(absl::ComputeCrc32c(
+            {reinterpret_cast<char const*>(_buffer.data() + startPos), size})),
+        .size = static_cast<std::uint32_t>(size + sizeof(Record::Footer))};
+    TRI_ASSERT(footer.size % 8 == 0);
+    _buffer.append(footer);
+  }
+
+  Buffer _buffer;
+};
+
+}  // namespace
+
+LogPersistor::LogPersistor(LogId logId,
+                           std::shared_ptr<IFileManager> fileManager)
+    : _logId(logId), _fileManager(std::move(fileManager)) {
+  LOG_TOPIC("a5ceb", TRACE, Logger::REPLICATED_WAL)
+      << "Creating LogPersistor for log " << logId;
+
+  // TODO - implement segmented logs
+  auto filename = std::to_string(logId.id()) + ".log";
+  _fileSet.emplace(LogFile{.filename = filename});
+  _activeFile = _fileManager->createWriter(filename);
+  auto fileReader = _activeFile->getReader();
+  if (fileReader->size() == 0) {
+    FileHeader header = {.magic = wMagicFileType, .version = wCurrentVersion};
+    auto res = _activeFile->append(header);
```

Comment:
No, I want to keep the FileManager, FileReader and FileWriter completely plain and independent of any log specific details. However, it would probably make sense to introduce a LogWriter similar to the LogReader.